src/deploy.ts

```diff
@@ -1,209 +1,378 @@
+    );
+    return router;
+  }
+
+  protected async deploySynthetic(
+    chain: ChainName,
+    config: HypERC721Config,
+  ): Promise<HypERC721> {
+    let router: HypERC721;
+    if (isUriConfig(config)) {
+      router = await this.deployContractFromFactory(
         chain,
-        isUriConfig(config)
-          ? new HypERC721URIStorage__factory()
-          : new HypERC721__factory(),
-        `HypERC721${isUriConfig(config) ? 'URIStorage' : ''}`,
+        new HypERC721URIStorage__factory(),
+        'HypERC721URIStorage',
         [],
       );
-      await this.multiProvider.handleTx(
+    } else {
+      router = await this.deployContractFromFactory(
         chain,
-        router.initialize(
-          config.mailbox,
-          config.interchainGasPaymaster,
-          config.totalSupply,
-          config.name,
-          config.symbol,
-        ),
+        new HypERC721__factory(),
+        'HypERC721',
+        [],
       );
-      return { router };
     }
-    throw new Error('Invalid config');
+    await this.multiProvider.handleTx(
+      chain,
+      router.initialize(
+        config.mailbox,
+        config.interchainGasPaymaster,
+        config.totalSupply,
+        config.name,
+        config.symbol,
+      ),
+    );
+    return router;
+  }
+
+  router(contracts: HyperlaneContracts<HypERC721Factories>) {
+    return contracts.router;
+  }
+
+  async deployContracts(chain: ChainName, config: HypERC721Config) {
+    let router: HypERC721 | HypERC721Collateral;
+    if (isCollateralConfig(config)) {
+      router = await this.deployCollateral(chain, config);
+    } else if (isSyntheticConfig(config)) {
+      router = await this.deploySynthetic(chain, config);
+    } else {
+      throw new Error('Invalid ERC721 token router config');
+    }
+    return { router };
+  }
+
+  async buildTokenMetadata(
+    configMap: ChainMap<TokenConfig>,
+  ): Promise<ChainMap<TokenMetadata>> {
+    let tokenMetadata: TokenMetadata | undefined;
+
+    for (const [chain, config] of Object.entries(configMap)) {
+      if (isCollateralConfig(config)) {
+        const collateralMetadata = await HypERC721Deployer.fetchMetadata(
+          this.multiProvider.getProvider(chain),
+          config,
+        );
+        tokenMetadata = {
+          ...collateralMetadata,
+          totalSupply: 0,
+        };
+      } else if (isTokenMetadata(config)) {
+        tokenMetadata = config;
+      }
+    }
+
+    if (!isTokenMetadata(tokenMetadata)) {
+      throw new Error('Invalid ERC721 token metadata');
+    }
+
+    return objMap(configMap, () => (tokenMetadata!));
+  }
+
+  buildGasOverhead(configMap: ChainMap<TokenConfig>): ChainMap<GasConfig> {
+    return objMap(configMap, (_, config) => ({
+      gas: HypERC721Deployer.gasOverheadDefault(config),
+    }));
+  }
+
+  async deploy(configMap: ChainMap<TokenConfig & RouterConfig>) {
+    const tokenMetadata = await this.buildTokenMetadata(configMap);
```

Comment:
This is obviously related to https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/2038 for when this throws an error as a check. If there are multiple checks, it would be nice to collect all checks that are failing, similar to the checker violations today. Probably outside the scope of this PR but something to note for @asaj when he works on this